src/Microsoft.AspNetCore.Server.IIS/Core/IO/AsyncIOOperation.cs

```diff
@@ -101,7 +103,7 @@ public AsyncContinuation Complete(int hr, int bytes)
                 _result = bytes;
                 if (hr != NativeMethods.HR_OK)
                 {
-                    _exception = new IOException("IO exception occurred", hr);
+                    _exception = hr != NativeMethods.COR_E_IO ? new IOException("Native IO operation failed") : Marshal.GetExceptionForHR(hr);
```

Comment:
We may want to consider wrapping all GetExcecptionForHR in a custom exception (like IISException) as COM exceptions are not very discoverable.